lib/Conversion/CalyxToHW/CalyxToHW.cpp

```diff
@@ -123,23 +123,30 @@ struct ConvertAssignOp : public OpConversionPattern<calyx::AssignOp> {
   LogicalResult
   matchAndRewrite(calyx::AssignOp assign, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
-    Value dest = adaptor.getDest();
+    Value src = adaptor.getSrc();
+    if (auto guard = adaptor.getGuard()) {
+      auto zero =
+          rewriter.create<hw::ConstantOp>(assign.getLoc(), src.getType(), 0);
+      src = rewriter.create<MuxOp>(assign.getLoc(), guard, src, zero);
+      for (Operation *destUser : assign.getDest().getUsers()) {
+        if (destUser == assign)
+          continue;
+        if (auto otherAssign = dyn_cast<calyx::AssignOp>(destUser)) {
+          src = rewriter.create<MuxOp>(assign.getLoc(), otherAssign.getGuard(),
+                                       otherAssign.getSrc(), src);
+          rewriter.eraseOp(destUser);
```

Comment:
Can you avoid this if-condition by having the `src` initially be 0?

```c++
auto src =
    rewriter.create<hw::ConstantOp>(assign.getLoc(), src.getType(), 0);
for (Operation *destUser : assign.getDest().getUsers()) {
    if (auto otherAssign = dyn_cast<calyx::AssignOp>(destUser)) {
        src = rewriter.create<MuxOp>(assign.getLoc(), otherAssign.getGuard(),
                                    otherAssign.getSrc(), src);
        rewriter.eraseOp(destUser);
    }
}
```

Reply:
Makes sense. Since there's no SSA values being replaced by the `rewriter.replaceOpWithNewOp<sv::AssignOp>(assign, dest, src);`, would it make sense to then just do `rewriter.create<sv::AssignOp>(/*assign.getLoc() stored earlier*/ loc, dest, src);` instead?

Reply:
In the case of no guards (just a single `calyx.assign`), `rewriter.create<sv::AssignOp>` wll be called without the `calyx.assign` being erased. Will it replace it (since we'll pass the same `loc`) or add it before/after it?

Reply:
Another way to do it:
```cpp
  LogicalResult
  matchAndRewrite(calyx::AssignOp assign, OpAdaptor adaptor,
                  ConversionPatternRewriter &rewriter) const override {
    // To make life easy in ConvertComponentOp, we read from the output wires so
    // the dialect conversion block argument mapping would work without a type
    // converter. This means assigns to ComponentOp outputs will try to assign
    // to a read from a wire, so we need to map to the wire.
    Value dest = adaptor.getDest();
    if (auto readInOut = dyn_cast<ReadInOutOp>(dest.getDefiningOp()))
      dest = readInOut.getInput();

    Value src = adaptor.getSrc();
    if (auto guard = adaptor.getGuard()) {
      Location loc = assign.getLoc();
      src = rewriter.create<hw::ConstantOp>(assign.getLoc(), src.getType(), 0);
      for (Operation *destUser :
           llvm::make_early_inc_range(assign.getDest().getUsers())) {
        if (auto otherAssign = dyn_cast<calyx::AssignOp>(destUser)) {
          src = rewriter.create<MuxOp>(assign.getLoc(), otherAssign.getGuard(),
                                       otherAssign.getSrc(), src);
          rewriter.eraseOp(destUser);
        }
      }
      rewriter.create<sv::AssignOp>(loc, dest, src);
    } else {
      rewriter.replaceOpWithNewOp<sv::AssignOp>(assign, dest, src);
    }
    return success();
  }
  ```

Reply:
... i'll concede, let's keep the original :).